src/governance/classes.cpp

```diff
@@ -495,16 +495,29 @@ CAmount CSuperblock::GetPaymentsLimit(int nBlockHeight)
     }
 
     const CBlockIndex* tipIndex = ::ChainActive().Tip();
-    bool fMNRewardReallocated = llmq::utils::IsMNRewardReallocationActive(tipIndex);
-    bool fV20Active = llmq::utils::IsV20Active(tipIndex);
+
+    const auto v20_state = llmq::utils::GetV20State(tipIndex);
+    bool fV20Active{v20_state == ThresholdState::ACTIVE};
+    if (!fV20Active && nBlockHeight > tipIndex->nHeight) {
+        // If fV20Active isn't active yet and nBlockHeight refers to a future SuperBlock
+        // then we need to check if the fork is locked_in and see if it will be active by the time of the future SuperBlock
+        if (v20_state == ThresholdState::LOCKED_IN) {
+            int activation_height = llmq::utils::GetV20Since(tipIndex) + static_cast<int>(Params().GetConsensus().vDeployments[Consensus::DEPLOYMENT_V20].nWindowSize);
+            if (nBlockHeight >= activation_height) {
+                fV20Active = true;
+            }
+        }
+    }
+
+    const auto mn_rr_state = llmq::utils::GetMNRewardReallocationState(tipIndex);
+    bool fMNRewardReallocated{mn_rr_state == ThresholdState::ACTIVE};
     if (!fMNRewardReallocated && nBlockHeight > tipIndex->nHeight) {
         // If fMNRewardReallocated isn't active yet and nBlockHeight refers to a future SuperBlock
         // then we need to check if the fork is locked_in and see if it will be active by the time of the future SuperBlock
-        if (llmq::utils::GetMNRewardReallocationState(tipIndex) == ThresholdState::LOCKED_IN) {
+        if (mn_rr_state == ThresholdState::LOCKED_IN) {
```

Comment:
can it somehow cause a hard-fork to happen before `mn_rr` state is in `ACTIVE` state?

does this check changes consensus rules?

Reply:
I'm not sure I understand your question... can you clarify it pls?

Reply:
I am just wondering:
 `GetPaymentsLimits()` is used in `IsBlockValueValid()` in masternode/payments.cpp that is called from `CChainState::ConnectBlock`

So if it would be different value that is used in calculation at the block at height where the status is only LOCKED_IN but not ACTIVE, would not be there consensus rules (IsBlockValueValid) to be changed before status is changed to ACTIVE but just LOCKED_IN?

If so, may hard-fork happens before ACTIVE status but somewhere during LOCKED_IN?

Reply:
I don't know, I maybe missing smth... `ConnectBlock` works for the `tip+1` block height and this block of code in `GetPaymentsLimits` ensures it can handle all blocks up to `tip+nWindowSize` correctly. I don't see a code path that would lead to any inconsistencies or a potential hf here 🤷‍♂️